### PR TITLE
allow to compute euler number along axis

### PR DESCRIPTION
**Pull request by vindex10**
Hi All! Recently I found myself in a situation where I needed to compute euler number for billions of images. I usually do this in batches. But, it turned out that the current `euler_number` functionality doesn't support evaluation along axis of the numpy array.

Mainly, this limitation is coming from `np.histogram`, specifically `np.bincount` that doesn't support evaluation along axis.

In my practice, the commonly used alternative is dask-backed [xhistogram](https://github.com/xgcm/xhistogram). Below you can see the comparison in runtime between python for loop and xhistogram approaches (serial uses np.histogram):

```
n serial(python for loop) xhistogram
1 0.00012839853006880731 0.0003150630200980231
10 0.0012457112300035078 0.0003546601699781604
50 0.0057103785099752715 0.0005381384199426975
200 0.02288406733001466 0.0015507272200193257
800 0.12443387749997782 0.00789261197001906
3000 0.4918516820800141 0.030045623120095115
```

the difference is noticeable already for batches bigger than 10 images. 

I'm not sure gaining xhistogram/dask as a dependency for scikit-image is feasable. But let me know if there is a place for this solution in the scikit-image repo :)

### CODE REVIEW

**Comment by vindex10** [comment]
a code snippet I used to compare runtimes

<details>
<summary>Spoiler</summary>
<blockquote><pre>
"""
1 0.00012839853006880731 0.0003150630200980231
10 0.0012457112300035078 0.0003546601699781604
50 0.0057103785099752715 0.0005381384199426975
200 0.02288406733001466 0.0015507272200193257
800 0.12443387749997782 0.00789261197001906
3000 0.4918516820800141 0.030045623120095115
"""


import numpy as np
from skimage.measure import euler_number
from timeit import timeit


def main():
    test(1)
    test(10)
    test(50)
    test(200)
    test(800)
    test(3000)


SAMPLE = np.array(
    [[0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0],
     [0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0],
     [0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0],
     [0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0],
     [0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0],
     [0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0],
     [0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0],
     [1, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0],
     [0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1],
     [0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1]]
)


def test(n):
    data = np.array([SAMPLE for _ in range(n)])
    serial_hist_bench = bench(lambda: serial_test(data))
    x_hist_bench = bench(lambda: xhist_test(data))
    print(n, serial_hist_bench, x_hist_bench)


def bench(f):
    number = 100
    return timeit(f, number=number)/number


def serial_test(data):
    res = np.zeros(data.shape[0])
    for i, row in enumerate(data):
        res[i] = euler_number(row)
    return res


def xhist_test(data):
    return euler_number(data, axes=[1, 2])


if __name__ == "__main__":
    main()
</pre></blockquote>
</details>


I also add if in the code, to fallback to the numpy histogram:

```
    if axes is None:
        h = np.histogram(XF_flat, bins=bin_edges)[0]
    else:
        h = xh.histogram(XF_flat, bins=bin_edges, axis=-1)[0]
```

without it, using xhistogram when the input is 1 image, looks as follows:

```
1 0.00019371783011592924 0.00031000096001662316
10 0.0018349548400146886 0.00035402936002356
50 0.008764909289893694 0.0005373617800069042
200 0.03525784356999793 0.001865086629986763
800 0.1792115688699414 0.008241007879987591
3000 0.8140393774700351 0.031148748900013744
```

so, it looks like for 1 image it is better to use np.histogram. so I include this "if" into the pull request

**Comment by vindex10** [comment]
wow. it looks like dask is not a restricted trick :)
https://github.com/scikit-image/scikit-image/pull/5587/

so maybe I this one is not that heavy?

**Comment by grlee77** [comment]
Can you use `numpy.histogramdd`, with `bins` set to a tuple to achieve the same result? Specifically, I mean that one would specify only 1 bin along any axis we don't want to include. That seems like it might work and would avoid having to require any new dependencies.

If we did add `xhistogram` it would be in `optional.txt` rather than `default.txt` (as `dask[array]` currently is).

I think both `boost-histogram` and `dask-histogram` offer a numpy-like API, so hopefully we could potentially swap out the histogram implementation later in a follow-up PR if optional dependencies are present. There was also some prior discussion of faster histogram options in #5935,

**Comment by vindex10** [comment]
I was looking at histogramdd before. I'm not sure it will work.

the input array for it has shape (N, D). the output histogram will have the dimension D1*D2*D3... where Di - number of bins along each dimension. so basically, we will be interested in some partial sums along axes of the D-dimensional cube.

Maybe I don't see some nice shortcut here, but I think it would have been shared in these posts:
https://stackoverflow.com/questions/40591754/vectorizing-numpy-bincount
https://stackoverflow.com/questions/40588403/vectorized-searchsorted-numpy

Because, both of them are related to a vectorized histogramming in numpy

I took the references from the source code of xhistogram: https://xhistogram.readthedocs.io/en/latest/_modules/xhistogram/core.html#histogram

I must admit, I didn't check the source before I got to use it, I've got conviced by the benchmark pretty quickly. But it turned out, that xhistogram has just put those two stackoverflow posts above into a package. Dask turned out ot be an optional dependency.

Do you think it worth it to copy the implementation of the xhistogram to scikit image, or the lightweight numpy only dependency is fine?

**Comment by vindex10** [comment]
Hi! Just to revive the discussion :)

I had a look at boost-histogram, fast-histogram and I didn't find the obvious way there to apply histogram along 1 numpy axis efficiently.

At the same time, I think it would be possible to re-use xhistogram implementation that is purely numpy based:
https://github.com/xgcm/xhistogram/pull/75

I opened a pull request there. That's one of the ways. Otherwise, we can just borrow the implementation. it is 500 lines of code.


Just to emphasize, my goal is not to makethe computation of the euler-number faster itself, but to make the loop over images faster (when we need to compute euler number for a batch of images stored in one numpy array)

File: skimage/measure/_regionprops_utils.py [hunk 1]
```diff
@@ -174,13 +181,53 @@ def euler_number(image, connectivity=None):
     # XF has values in the 0-255 range in 3D, and in the 0-15 range in 2D,
     # with one unique value for each binary configuration of the
     # 27-voxel cube in 3D / 8-pixel square in 2D, up to symmetries
-    XF = ndi.convolve(image, config, mode='constant', cval=0)
-    h = np.bincount(XF.ravel(), minlength=bins)
+    XF = convolve_along_axes(image, config, image_axes, mode='constant', cval=0)
+    XF_flat = flatten_along_axes(XF, axes=image_axes)
+
+    bin_edges = np.arange(0, bins+0.1, 1)
+    h = histogram_along_axes(XF_flat, bins=bin_edges, axes=-1 if coord_dims else None)
```
Comment (grlee77) [inline skimage/measure/_regionprops_utils.py:188]: I think the following is equivalent and avoids the need for the dependency. It also runs a bit faster on the case I tried it on (see example below)

```suggestion
    h = np.empty((XF_flat.shape[0], bins), dtype=int)
    for i, x in enumerate(XF_flat):
        h[i, :] = np.bincount(x, minlength=bins)
```

I tried this using a simple benchmark:
```Python
import numpy as np
many_samples = np.ones((256, 512, 512), dtype=bool)
many_samples[:, ::16, :] = 0
many_samples[:, :, ::16] = 0
euler_number(many_samples, axes=[1, 2]
```
and the duration after the suggested change was 1.05 s vs. 1.88 s prior to the change.

**Comment by vindex10** [comment]
Hi @grlee77, thank you for you reply and suggestion. Indeed, it turned out that bincount in the for loop works faster in comparison to the xhistogram approach.

I noticed a small but important piece of constructing the histogram. Before doing bincount, `searchsorted` is required.

First of all I benchmarked my xhistogram implementation as a baseline, on your example:

```
bin_edges = np.arange(0, bins+0.1, 1)
h = histogram_along_axes(XF_flat, bins=bin_edges, axes=-1 if coord_dims else None)
```

2.687926361399968

Then I also ran your solution:

```
    XF_flat = flatten_along_axes(XF, axes=image_axes)
    h = np.empty((XF_flat.shape[0], bins), dtype=int)
    for i, x in enumerate(XF_flat):
        h[i, :] = np.bincount(x, minlength=bins)
```

1.4685133945000417

It was just to cross check. Indeed, around 80% faster.

Now we restore searchsorted:

```
    XF = np.searchsorted(bin_edges, XF)

    XF_flat = flatten_along_axes(XF, axes=image_axes)
    h = np.empty((XF_flat.shape[0], bins), dtype=int)
    for i, x in enumerate(XF_flat):
        h[i, :] = np.bincount(x, minlength=bins)
```

2.4116079455000543

And we get to ~8% boost.


I also tried to use the xhistogram inspired workaround with bincount, that repeated the result of the xhistogram itself:

```
    XF = np.searchsorted(bin_edges, XF)
    h = np.bincount((XF + bins * np.arange(XF.shape[0])[:, None, None]).ravel()).reshape((XF.shape[0], bins))
```

2.644724493149988

**Comment by vindex10** [comment]
I understood. In the original implementation there was no searchsorted. And it is indeed redundand!

**Comment by vindex10** [comment]
I did few more tests.

1. It looks like my initial benchmarking of the histograms was successful because of the low performance of the searchsorted (that we don't need here)
2. bincount trick is slower than just pythonic for loop
3. I ran the baseline (from main branch) and a resulting version (with for loop inside euler_number, without searchsorted).

for-loop outside euler_number, main branch: 1.46s
for loop inside euler number, wrapping bincount (this branch, basically solution suggested by @grlee77 ): 1.75s

Below I also attach flamegraphs for both baseline and a solution that resulted from the discussion
[res.svg.gz](https://github.com/scikit-image/scikit-image/files/7896785/res.svg.gz)
[base.svg.gz](https://github.com/scikit-image/scikit-image/files/7896786/base.svg.gz)

also a piece of code used for the benchmarking:
[test.py.gz](https://github.com/scikit-image/scikit-image/files/7896795/test.py.gz)


So, probably the PR can be closed. I'd like to thank you for the discussion! and hopefully this small investigation will save time for someone thinking of the same thing in future :)

**Comment by grlee77** [comment]
Thanks for doing further testing @vindex10, I will go ahead and close this.

**Pull request closed without merging**